### PR TITLE
Check transform shape for weight learning rules

### LINKED ISSUE #639: Unsupervised learning without initial transform breaks
```python
import nengo

with nengo.Network() as model:
    a = nengo.Ensemble(100, 1)
    b = nengo.Ensemble(100, 1)
    conn = nengo.Connection(a.neurons, b.neurons)
    conn.learning_rule_type = nengo.BCM()

sim = nengo.Simulator(model)
```
This breaks at build time because we try to cheat with the transform, since in the case of a normal connection we would just leave the transform as `np.array(1)` and take advantage of Numpy's broadcasting to apply it properly. In the case of a learned connection, though, the transform ends up too small, and the builder can't apply the learning rule update.

### PR DESCRIPTION
**Pull request by hunse**
Learning rules that modify 'weights' need a 2D transform or weight
solver. This checks that the transform is the right shape. Fixes #639.

### CODE REVIEW

**Comment by tbekolay** [comment]
This looks great to me! @drasmuss can you review?

File: nengo/connection.py [hunk 1]
```diff
@@ -59,12 +59,25 @@ def validate_rule(self, conn, rule):
...
+                raise ValueError(
+                    "Learning rule '%s' can not be applied to decoded "
+                    "connections. Try setting solver.weights to True or "
+                    "connecting between two Neurons objects." % rule)
+
+            # transform matrix must be 2D
+            pre_size = (
+                conn.pre_obj.n_neurons if isinstance(conn.pre_obj, Ensemble)
+                else conn.pre.size_out)
+            post_size = conn.post.size_out
```
Comment (drasmuss) [inline nengo/connection.py]: Should this be `conn.post.size_in`?

**Comment by drasmuss** [comment]
Other than that one thing, looks good!

File: nengo/connection.py (hunk 1 above)
Comment (hunse) [inline nengo/connection.py]: Hmm, yes, very good point!

**Comment by hunse** [comment]
Ok, fixed that. Thanks for catching it, @drasmuss!

**Pull request merged**